core/controllers/reader.py

```diff
@@ -122,31 +123,67 @@ def classify(exp_id, state, answer, params):
-            best_matched_truth_value = ored_truth_value
-            best_matched_rule_spec = best_rule_spec
-            best_matched_answer_group = answer_group
-            best_matched_answer_group_index = i
+            if rule_spec.rule_type != rule_domain.FUZZY_RULE_TYPE:
+                evaluated_truth_value = rule_domain.evaluate_rule(
+                    rule_spec, input_type, params, normalized_answer, fs)
+                if evaluated_truth_value > best_matched_truth_value:
+                    best_matched_truth_value = evaluated_truth_value
+                    best_matched_rule_spec = rule_spec
+                    best_matched_answer_group = answer_group
+                    best_matched_answer_group_index = i
+
+    # Find the maximum soft rule that matches. This is
+    # done by ORing (maximizing) all truth values of all rules over all answer
+    # groups. The group with the highest truth value is considered the best
+    # match.
+    if best_matched_answer_group is None:
+        for (i, answer_group) in enumerate(state.interaction.answer_groups):
+            for rule_spec in answer_group.rule_specs:
+                if rule_spec.rule_type == rule_domain.FUZZY_RULE_TYPE:
+                    evaluated_truth_value = rule_domain.evaluate_rule(
+                        rule_spec, input_type, params, normalized_answer, fs)
+                    if evaluated_truth_value > best_matched_truth_value:
+                        best_matched_truth_value = evaluated_truth_value
+                        best_matched_rule_spec = rule_spec
+                        best_matched_answer_group = answer_group
+                        best_matched_answer_group_index = i
+
+    # Run the classifier
+    if best_matched_answer_group is None:
+        sc = classifier_services.StringClassifier()
+        training_examples = [[doc, []] for doc in
+            state.interaction.confirmed_unclassified_answers]
+        for (i, answer_group) in enumerate(state.interaction.answer_groups):
+            for rule_spec in answer_group.rule_specs:
+                if rule_spec.rule_type == rule_domain.FUZZY_RULE_TYPE:
+                    training_examples.extend([[doc, [str(i)]] for doc in
+                        rule_spec.inputs['training_data']])
+        if len(training_examples) > 0:
+            sc.load_examples(training_examples)
+            doc_ids = sc.add_examples_for_predicting([answer])
+            predicted_label = sc.predict_label_for_doc(doc_ids[0])
+            if predicted_label != '_default':
+                predicted_answer_group_index = int(predicted_label)
+                predicted_answer_group = state.interaction.answer_groups[
+                    predicted_answer_group_index]
+                best_matched_truth_value = rule_domain.CERTAIN_TRUE_VALUE
+                for rule_spec in answer_group.rule_specs:
+                    if rule_spec.rule_type == rule_domain.FUZZY_RULE_TYPE:
```

Comment:
Your assumption is correct (that there *should* be only one fuzzy rule per answer group), but it is not actually guarded against in the code base. My concern above is defending against potentially undefined behavior if future Oppia changes lead to multiple fuzzy rules being placed in an answer group. I think it's good to assume one fuzzy rule per group, but we should make this explicit in case someone tries adding more than one in the future. Thus, I propose:

1. Defend against >1 fuzzy rules in an answer group by updating the validate() function in exp_domain.AnswerGroup.

1. Implement a corresponding test in exp_domain_test to verify the validation change works in the correct circumstances.

1. Update your loops in classify() to be consistent when they reach a fuzzy rule. One potential way to do this that might be rather clean is to implement a property in AnswerGroup that returns the single fuzzy rule in the group, or None if it does not have a fuzzy rule. This will simplify your construction of training data as well as selecting the fuzzy rule of an answer group chosen by the classifier.

If we should discuss anything above in greater detail, please let me know. I'll be happy to send some code samples or setup a quick meeting.